src/Microsoft.AspNet.Mvc.Razor/Razor/PreCompileViews/RazorPreCompiler.cs

```diff
@@ -58,93 +64,127 @@ public virtual void CompileViews([NotNull] IBeforeCompileContext context)
+                {
+                    cacheSetContext.AddExpirationTrigger(_fileSystem.Watch(viewStartPath));
+                }
             }
 
+            return entry;
+        }
+
+        private void GetFileInfosRecursive(string root, List<RelativeFileInfo> razorFiles)
+        {
+            var fileInfos = _fileSystem.GetDirectoryContents(root);
+
             foreach (var fileInfo in fileInfos)
             {
                 if (fileInfo.IsDirectory)
                 {
-                    var subPath = Path.Combine(path, fileInfo.Name);
-
-                    foreach (var info in GetFileInfosRecursive(subPath))
-                    {
-                        yield return info;
-                    }
+                    var subPath = Path.Combine(root, fileInfo.Name);
+                    GetFileInfosRecursive(subPath, razorFiles);
                 }
                 else if (Path.GetExtension(fileInfo.Name)
                          .Equals(FileExtension, StringComparison.OrdinalIgnoreCase))
                 {
-                    var relativePath = Path.Combine(currentPath, fileInfo.Name);
+                    var relativePath = Path.Combine(root, fileInfo.Name);
                     var info = new RelativeFileInfo(fileInfo, relativePath);
-                    yield return info;
+                    razorFiles.Add(info);
                 }
             }
         }
 
-        protected virtual RazorFileInfo ParseView([NotNull] RelativeFileInfo fileInfo,
-                                                  [NotNull] IBeforeCompileContext context)
+        protected virtual PrecompilationCacheEntry GetCacheEntry([NotNull] RelativeFileInfo fileInfo)
         {
             using (var stream = fileInfo.FileInfo.CreateReadStream())
             {
-                var results = _host.GenerateCode(fileInfo.RelativePath, stream);
+                var host = GetRazorHost();
+                var results = host.GenerateCode(fileInfo.RelativePath, stream);
 
-                foreach (var parserError in results.ParserErrors)
+                if (results.Success)
                 {
-                    var diagnostic = parserError.ToDiagnostics(fileInfo.FileInfo.PhysicalPath);
-                    context.Diagnostics.Add(diagnostic);
-                }
-
-                var generatedCode = results.GeneratedCode;
-
-                if (generatedCode != null)
-                {
-                    var syntaxTree = SyntaxTreeGenerator.Generate(generatedCode,
+                    var syntaxTree = SyntaxTreeGenerator.Generate(results.GeneratedCode,
                                                                   fileInfo.FileInfo.PhysicalPath,
                                                                   CompilationSettings);
-                    var fullTypeName = results.GetMainClassName(_host, syntaxTree);
+                    var fullTypeName = results.GetMainClassName(host, syntaxTree);
 
                     if (fullTypeName != null)
```

Comment:
... because you've added a couple more `null` checks for the return value of this method and its callers.